### PR TITLE
feat(shorebird_cli): improve output for gradle "unsupported file major version" errors

### LINKED ISSUE #1821: fix: `Unsupported class file major version 65` when running `shorebird init` on Linux
Reported on Discord ([chat link](https://discord.com/channels/1030243211995791380/1221828343746531358/1221832057118920906))

Output:

```
shorebird init -v                                                                                                         ✔ 
[Process.run] /home/kingwill101/code/linkboard/frontend/android/gradlew app:tasks --all --console=auto (in /home/kingwill101/code/linkboard/frontend/android)
⠼ Detecting product flavors... (3.5s)Exited with code 1

stdout:
> Task :gradle:compileJava NO-SOURCE
> Task :gradle:compileGroovy FAILED
1 actionable task: 1 executed


stderr:

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/home/kingwill101/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s

✗ Detecting product flavors (3.6s)
Unable to extract product flavors.
Exception: > Task :gradle:compileJava NO-SOURCE
> Task :gradle:compileGroovy FAILED
1 actionable task: 1 executed


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/home/kingwill101/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s
```

### PR DESCRIPTION
**Pull request by erickzanardo**
## Description

This PR adds a common error handling system in the gradlew executable class and adds a handler for the `Unsupported class file major version`, which will link the user to our trouble shooting section.

This PR introduces some new concepts to the CLI so it is mostly a proposal, none of the technical choices here are strong opinionated and I am happy to change if there are better approaches!

Before this PR, this were the output of the CLI when it broken:

```
⠼ Detecting product flavors... (3.5s)Exited with code 1

stdout:
> Task :gradle:compileJava NO-SOURCE
> Task :gradle:compileGroovy FAILED
1 actionable task: 1 executed


stderr:

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/home/kingwill101/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s

✗ Detecting product flavors (3.6s)
Unable to extract product flavors.
Exception: > Task :gradle:compileJava NO-SOURCE
> Task :gradle:compileGroovy FAILED
1 actionable task: 1 executed


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':gradle:compileGroovy'.
> BUG! exception in phase 'semantic analysis' in source unit '/home/kingwill101/flutter/packages/flutter_tools/gradle/src/main/groovy/app_plugin_loader.groovy' Unsupported class file major version 65

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 3s
```

After the PR, this is the output
Screenshot 2024-06-18 at 17 47 22

Fixes #1821 

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore
- [ ] 🧪 Tests

### CODE REVIEW

File: packages/shorebird_cli/lib/src/executables/gradlew.dart [hunk 1]
```diff
@@ -82,6 +134,16 @@ class Gradlew {
         if (!javaHome.isNullOrEmpty) 'JAVA_HOME': javaHome!,
       },
     );
+
+    if (result.exitCode != ExitCode.success.code) {
+      for (final error in GradleHandedErrors.values) {
+        if (error.apply(result)) {
+          throw error.toException();
+        }
+      }
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/executables/gradlew.dart]: This is a bit complex imo. I'd probably just keep it simple like:

```dart
if (error.contains('Unsupported class file major version')) {
  throw IncompatibleGradleVersion();
}
```

And define `IncompatibleGradleVersion` like:

```dart
class IncompatibleGradleVersion extends GradleProcessException {
  IncompatibleGradleVersion() : super(message: '<INCLUDE MESSAGE HERE>');
}
```

File: packages/shorebird_cli/lib/src/executables/gradlew.dart [hunk 2]
```diff
@@ -3,13 +3,62 @@
...
+/// {@endtemplate}
+class GradleProcessException implements Exception {
+  /// {@macro gradle_process_exception}
+  const GradleProcessException(this.message);
+
+  final String message;
+
+  @override
+  String toString() {
+    return 'Gradle sub-process failed with error:\n$message';
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/executables/gradlew.dart]:
```suggestion
    return 'Gradle error:\n$message';
```

File: packages/shorebird_cli/lib/src/executables/gradlew.dart (hunk 1 above)
Comment (erickzanardo) [reply packages/shorebird_cli/lib/src/executables/gradlew.dart]: Sounds good, I made it simpler in a change inspired by your suggestion, I guess I did overcomplicated indeed.

File: packages/shorebird_cli/lib/src/executables/gradlew.dart [hunk 3]
```diff
@@ -82,6 +108,16 @@ class Gradlew {
         if (!javaHome.isNullOrEmpty) 'JAVA_HOME': javaHome!,
       },
     );
+
+    if (result.exitCode != ExitCode.success.code) {
+      if (result.stderr
+          .toString()
+          .contains(IncompatibleGradleException.errorPattern)) {
+        throw IncompatibleGradleException();
+      }
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/executables/gradlew.dart:117]: Shouldn't we also throw a generic gradle exception for other errors?

**Review by felangel — APPROVED** [review]

File: packages/shorebird_cli/lib/src/executables/gradlew.dart (hunk 3 above)
Comment (erickzanardo) [reply packages/shorebird_cli/lib/src/executables/gradlew.dart:117]: We don't need to, the result with the failed exit code returned (just like it was before) and the methods that call `_run` will handle the error the same way they would.

File: packages/shorebird_cli/lib/src/executables/gradlew.dart (hunk 1 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/executables/gradlew.dart]: All good, I like your latest change 👍

**Pull request merged**